extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java

```diff
@@ -942,6 +962,172 @@ public void testRestore() throws Exception
     Assert.assertEquals(ImmutableList.of("d", "e"), readSegmentDim1(desc2));
   }
 
+  @Test(timeout = 60_000L)
+  public void testRunWithPauseAndResume() throws Exception
+  {
+    final KafkaIndexTask task = createTask(
+        null,
+        new KafkaIOConfig(
+            "sequence0",
+            new KafkaPartitions("topic0", ImmutableMap.of(0, 2L)),
+            new KafkaPartitions("topic0", ImmutableMap.of(0, 5L)),
+            kafkaServer.consumerProperties(),
+            true,
+            false
+        ),
+        null
+    );
+
+    final ListenableFuture<TaskStatus> future = runTask(task);
+
+    // Insert some data, but not enough for the task to finish
+    try (final KafkaProducer<byte[], byte[]> kafkaProducer = kafkaServer.newProducer()) {
+      for (ProducerRecord<byte[], byte[]> record : Iterables.limit(RECORDS, 4)) {
+        kafkaProducer.send(record).get();
+      }
+    }
+
+    while (countEvents(task) != 2) {
+      Thread.sleep(25);
+    }
+
+    Assert.assertEquals(2, countEvents(task));
+    Assert.assertEquals(KafkaIndexTask.Status.READING, task.getStatus());
+
+    Map<Integer, Long> currentOffsets = objectMapper.readValue(
+        task.pause(0).getEntity().toString(),
+        new TypeReference<Map<Integer, Long>>()
+        {
+        }
+    );
+    Assert.assertEquals(KafkaIndexTask.Status.PAUSED, task.getStatus());
+
+    // Insert remaining data
+    try (final KafkaProducer<byte[], byte[]> kafkaProducer = kafkaServer.newProducer()) {
+      for (ProducerRecord<byte[], byte[]> record : Iterables.skip(RECORDS, 4)) {
+        kafkaProducer.send(record).get();
+      }
+    }
+
+    boolean timedOut = false;
+    try {
+      future.get(10, TimeUnit.SECONDS);
+    }
+    catch (TimeoutException e) {
+      timedOut = true;
+    }
+    Assert.assertTrue("Task completed when it should have been paused", timedOut);
```

Comment:
could avoid timeOut variable by moving this line after 1014 and change to Assert.fail("...") instead.